File: src/프로그래머스_Level2/문자열압축.js
function ReplaceAt(input, replace, start, end) {
    let tmp = [];
    for (let idx = 0; idx < replace.length; idx++) tmp.push(replace[idx]);

    let arr1 = input.slice(0, start);
    let arr2 = input.slice(end + 1);

    input.splice(start, end - start + 1);

    return arr1.concat(tmp, arr2);
}

function solution(s) {
    let answer = s.length;

    for (let delta = 1; delta < s.length; delta++) {
        let sb = [];

        let start = 0,
            end = 0 + delta;

        while (start != end && end <= s.length) {
            let splited = s.substring(start, end);

            if (sb.length == 0) {
                for (let idx = 0; idx < splited.length; idx++) {
                    sb.push(splited[idx]);
                }
            } else {
                let prev = sb.slice(sb.length - delta, sb.length).join("");

                if (prev.length != 0 && prev === splited) {
                    let countIndexEnd = sb.length - delta - 1;

                    if (countIndexEnd >= 0) {
                        if (!isNaN(sb[countIndexEnd])) {
                            let countIndexStart = countIndexEnd - 1;

                            while (countIndexStart >= 0 && !isNaN(sb[countIndexStart])) countIndexStart--;
                            countIndexStart++;

                            let count = parseInt(sb.slice(countIndexStart, countIndexEnd + 1).join(""));
                            sb = ReplaceAt(sb, count + 1 + "", countIndexStart, countIndexEnd);
                        } else {
                            sb.splice(countIndexEnd + 1, 0, "2");
                        }
                    } else {
                        sb.splice(0, 0, "2");
                    }
                } else {
                    for (let idx = 0; idx < splited.length; idx++) {
                        sb.push(splited[idx]);
                    }
                }
            }

            start = end;
            end = start + delta > s.length ? s.length : start + delta;
        }

        // console.log("delta : " + delta + ", sb : " + sb.join("") + ", sb.length : " + sb.length);
        answer = Math.min(answer, sb.length);
    }

    return answer;
}

// solution("aabbaccc");
// solution("ababcdcdababcdcd");
// solution("abcabcdede");
// solution("abcabcabcabcdededededede");
// solution("xababcdcdababcdcd");
solution("xxxxxxxxxxyyy");

File: src/프로그래머스_Level2/문자열 압축.py
def solution(s):
    if len(s) == 1:
        return 1

    answer = [] 
    
    for i in range(1, len(s) // 2 + 1):
        compressed = ""
        cnt = 1
        tmp = s[:i]

        for j in range(i, len(s), i):
            if tmp == s[j:i+j]:
                cnt += 1
            else:
                if cnt != 1:
                    compressed = compressed + str(cnt) + tmp
                else:
                    compressed = compressed + tmp

                tmp = s[j:i+j]
                cnt = 1

        if cnt != 1:
            compressed = compressed + str(cnt) + tmp
        else:
            compressed = compressed + tmp
        
        answer.append(len(compressed))

    return min(answer)

print(solution("aabbaccc"))
print(solution("ababcdcdababcdcd"))
print(solution("abcabcdede"))
print(solution("abcabcabcabcdededededede"))
print(solution("xababcdcdababcdcd"))
